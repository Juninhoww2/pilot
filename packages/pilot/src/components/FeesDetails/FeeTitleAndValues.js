import React from 'react'
import PropTypes from 'prop-types'
import { Flexbox } from 'former-kit'
import formatPercent from '../../formatters/percent'
import formatCurrency from '../../formatters/currency'
import styles from './styles.css'

const formatterByType = {
  currency: formatCurrency,
  percent: formatPercent,
}

const FeeTitleAndValues = ({ t, title, values }) => (
  <div>
    <p className={styles.title}>{title}</p>
    <Flexbox className={styles.feeValuesWrapper}>
      {values.map(({ fees, translationPath }) => (
        <div key={translationPath}>
          <p>{t(translationPath)}</p>
          <p>
            <strong>
              {
                fees
                  .map(({ type, value, valueSuffixPath }) => {
                    const formatter = formatterByType[type]
                    const formattedValue = formatter(value) || 'N/A'
                    const valueSuffix = valueSuffixPath ? t(valueSuffixPath) : ''

                    return `${formattedValue} ${valueSuffix}`
                  })
                  .join('+ ')
              }
            </strong>
          </p>
        </div>
      ))}
    </Flexbox>
  </div>
)

FeeTitleAndValues.propTypes = {
  t: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(PropTypes.shape({
    fees: PropTypes.arrayOf(PropTypes.shape({
      type: PropTypes.string.isRequired,
      value: PropTypes.number,
      valueSuffixPath: PropTypes.string,
    })),
    translationPath: PropTypes.string.isRequired,
  })).isRequired,
}

export default FeeTitleAndValues
