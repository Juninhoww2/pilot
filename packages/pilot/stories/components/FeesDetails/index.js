import React from 'react'
import { path, split } from 'ramda'
import Section from '../../Section'
import FeesDetails from '../../../src/components/FeesDetails'
import translations from '../../../public/locales/pt/translations.json'

const t = (sentence = '') => path(split('.', sentence), translations)

const FeesDetailsDefaultExample = () => (
  <Section>
    <FeesDetails
      fees={{
        anticipation: undefined,
        antifraud: 70,
        boleto: 380,
        gateway: undefined,
        installments: [
          { installment: 1, mdr: 1.2 },
          { installment: 2, mdr: 2.3 },
          { installment: 7, mdr: 3 },
        ],
        pix: {
          paymentFixedFee: 380,
          paymentSpreadFee: 1.19,
        },
        transfer: 187,
      }}
      t={t}
    />
  </Section>
)

const FeesDetailsMDRzaoExample = () => (
  <Section>
    <FeesDetails
      fees={{
        anticipation: 1.99,
        antifraud: 70,
        boleto: 380,
        gateway: undefined,
        installments: [
          { installment: 1, mdr: 1.2 },
        ],
        pix: {
          paymentFixedFee: 380,
          paymentSpreadFee: 1.19,
        },
        transfer: 187,
      }}
      isMDRzao
      t={t}
    />
  </Section>
)

export default {
  FeesDetailsDefault: FeesDetailsDefaultExample,
  FeesDetailsMDRzao: FeesDetailsMDRzaoExample,
}
