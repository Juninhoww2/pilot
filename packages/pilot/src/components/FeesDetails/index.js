import React from 'react'
import PropTypes from 'prop-types'
import { Flexbox } from 'former-kit'
import FeeTitleAndValues from './FeeTitleAndValues'
import styles from './styles.css'

const installmentsTranslations = {
  DEFAULT: {
    1: 'pages.empty_state.fees.one_installment',
    2: 'pages.empty_state.fees.two_to_six_installments',
    7: 'pages.empty_state.fees.seven_or_more_installments',
  },
  MDRZAO: {
    1: 'pages.empty_state.fees.mdrzao_installment',
  },
}

const buildInstallmentsValues = (
  anticipationType,
  installments
) => installments.map(item => ({
  translationPath: installmentsTranslations[anticipationType][item.installment],
  type: 'percent',
  value: item.mdr,
}))

const loadFirstInstallmentMDR = (installments) => {
  if (installments.length < 1) {
    return null
  }

  return installments[0].mdr
}

const buildCreditCardFees = ({ fees, isMDRzao }) => (
  isMDRzao
    ? [
      {
        translationPath: 'pages.empty_state.fees.mdrzao_installment',
        type: 'percent',
        value: loadFirstInstallmentMDR(fees.installments),
        valueSuffixPath: 'pages.empty_state.fees.per_transaction',
      },
    ]
    : [...buildInstallmentsValues('DEFAULT', fees.installments)])

const buildProcessingFees = (fees) => {
  const processingFees = [
    {
      translationPath: 'pages.empty_state.fees.processing',
      type: 'currency',
      value: fees.gateway,
    },
    {
      translationPath: 'pages.empty_state.fees.antifraud',
      type: 'currency',
      value: fees.antifraud,
    },
  ]

  return processingFees.filter(v => v.value !== 0)
}

const FeesDetails = ({ fees, isMDRzao, t }) => {
  const creditCardFees = buildCreditCardFees({ fees, isMDRzao })
  const processingFees = buildProcessingFees(fees)

  return (
    <div>
      <FeeTitleAndValues
        t={t}
        title={t('pages.empty_state.fees.credit_card')}
        values={[...creditCardFees, ...processingFees]}
      />
      <Flexbox className={styles.marginRight}>
        <FeeTitleAndValues
          t={t}
          title={t('pages.empty_state.fees.boleto')}
          values={[
            {
              translationPath: 'pages.empty_state.fees.paid',
              type: 'currency',
              value: fees.boleto,
            },
          ]}
        />
        <FeeTitleAndValues
          t={t}
          title={t('pages.empty_state.fees.anticipation')}
          values={[
            {
              translationPath: 'pages.empty_state.fees.tax',
              type: 'percent',
              value: fees.anticipation,
              valueSuffixPath: isMDRzao ? 'pages.empty_state.fees.per_installment' : '',
            },
          ]}
        />
        <FeeTitleAndValues
          t={t}
          title={t('pages.empty_state.fees.transfers')}
          values={[
            {
              translationPath: 'pages.empty_state.fees.doc_ted',
              type: 'currency',
              value: fees.transfer,
            },
          ]}
        />
      </Flexbox>
    </div>
  )
}

FeesDetails.propTypes = {
  fees: PropTypes.shape({
    anticipation: PropTypes.number,
    antifraud: PropTypes.number,
    boleto: PropTypes.number,
    gateway: PropTypes.number,
    installments: PropTypes.arrayOf(PropTypes.shape({
      installment: PropTypes.number.isRequired,
      mdr: PropTypes.number.isRequired,
    })),
    transfer: PropTypes.number,
  }),
  isMDRzao: PropTypes.bool,
  t: PropTypes.func.isRequired,
}

FeesDetails.defaultProps = {
  fees: {
    anticipation: null,
    antifraud: null,
    boleto: null,
    gateway: null,
    installments: [],
    transfer: null,
  },
  isMDRzao: false,
}

export default FeesDetails
