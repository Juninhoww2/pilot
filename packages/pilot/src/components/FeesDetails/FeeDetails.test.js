import React from 'react'
import { identity } from 'ramda'
import { render } from '@testing-library/react'
import FeesDetails from '.'

const feesBase = {
  anticipation: 3.14,
  antifraud: 70,
  boleto: 380,
  gateway: 50,
  pix: {
    paymentFixedFee: 380,
    paymentSpreadFee: 1.19,
  },
  transfer: 367,
}

test('render 1, 2 and 7 installments with all fees', () => {
  const fees = {
    ...feesBase,
    installments: [
      { installment: 1, mdr: 3.79 },
      { installment: 2, mdr: 4.19 },
      { installment: 7, mdr: 4.59 },
    ],
  }

  const {
    container,
    getByText,
    queryByText,
  } = render(<FeesDetails fees={fees} t={identity} />)

  getByText(/fees.one_installment/)
  getByText(/fees.two_to_six_installments/)
  getByText(/fees.seven_or_more_installments/)
  expect(queryByText(/fees.per_transaction/)).toBe(null)
  expect(container).toMatchSnapshot()
})

test('render 1 installments with all fees and isMDRzao enabled', () => {
  const fees = {
    ...feesBase,
    installments: [
      { installment: 1, mdr: 3.79 },
    ],
  }

  const {
    container,
    getByText,
  } = render(<FeesDetails fees={fees} t={identity} isMDRzao />)

  getByText(/fees.mdrzao_installment/)
  getByText(/fees.per_transaction/)

  expect(container).toMatchSnapshot()
})

test('render no installments are provided', () => {
  const fees = {
    ...feesBase,
    installments: [],
  }

  const {
    container,
    queryByText,
  } = render(<FeesDetails fees={fees} t={identity} />)

  expect(queryByText(/fees.one_installment/)).toBe(null)
  expect(queryByText(/fees.two_to_six_installments/)).toBe(null)
  expect(queryByText(/fees.seven_or_more_installments/)).toBe(null)
  expect(queryByText(/fees.per_transaction/)).toBe(null)
  expect(container).toMatchSnapshot()
})
