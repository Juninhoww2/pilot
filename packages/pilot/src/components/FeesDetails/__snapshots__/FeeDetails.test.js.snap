// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1 installments with all fees and isMDRzao enabled 1`] = `
<div>
  <div>
    <div>
      <p
        class="title"
      >
        pages.empty_state.fees.credit_card
      </p>
      <div
        class="feeValuesWrapper"
      >
        <div>
          <p>
            pages.empty_state.fees.mdrzao_installment
          </p>
          <p>
            <strong>
              3,79%
               
              pages.empty_state.fees.per_transaction
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.processing
          </p>
          <p>
            <strong>
              R$ 0,50
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.antifraud
          </p>
          <p>
            <strong>
              R$ 0,70
               
            </strong>
          </p>
        </div>
      </div>
    </div>
    <div
      class="marginRight"
    >
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.boleto
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.paid
            </p>
            <p>
              <strong>
                R$ 3,80
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.anticipation
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.tax
            </p>
            <p>
              <strong>
                3,14%
                 
                pages.empty_state.fees.per_installment
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.transfers
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.doc_ted
            </p>
            <p>
              <strong>
                R$ 3,67
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render 1, 2 and 7 installments with all fees 1`] = `
<div>
  <div>
    <div>
      <p
        class="title"
      >
        pages.empty_state.fees.credit_card
      </p>
      <div
        class="feeValuesWrapper"
      >
        <div>
          <p>
            pages.empty_state.fees.one_installment
          </p>
          <p>
            <strong>
              3,79%
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.two_to_six_installments
          </p>
          <p>
            <strong>
              4,19%
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.seven_or_more_installments
          </p>
          <p>
            <strong>
              4,59%
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.processing
          </p>
          <p>
            <strong>
              R$ 0,50
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.antifraud
          </p>
          <p>
            <strong>
              R$ 0,70
               
            </strong>
          </p>
        </div>
      </div>
    </div>
    <div
      class="marginRight"
    >
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.boleto
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.paid
            </p>
            <p>
              <strong>
                R$ 3,80
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.anticipation
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.tax
            </p>
            <p>
              <strong>
                3,14%
                 
                
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.transfers
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.doc_ted
            </p>
            <p>
              <strong>
                R$ 3,67
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render no installments are provided 1`] = `
<div>
  <div>
    <div>
      <p
        class="title"
      >
        pages.empty_state.fees.credit_card
      </p>
      <div
        class="feeValuesWrapper"
      >
        <div>
          <p>
            pages.empty_state.fees.processing
          </p>
          <p>
            <strong>
              R$ 0,50
               
            </strong>
          </p>
        </div>
        <div>
          <p>
            pages.empty_state.fees.antifraud
          </p>
          <p>
            <strong>
              R$ 0,70
               
            </strong>
          </p>
        </div>
      </div>
    </div>
    <div
      class="marginRight"
    >
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.boleto
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.paid
            </p>
            <p>
              <strong>
                R$ 3,80
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.anticipation
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.tax
            </p>
            <p>
              <strong>
                3,14%
                 
                
              </strong>
            </p>
          </div>
        </div>
      </div>
      <div>
        <p
          class="title"
        >
          pages.empty_state.fees.transfers
        </p>
        <div
          class="feeValuesWrapper"
        >
          <div>
            <p>
              pages.empty_state.fees.doc_ted
            </p>
            <p>
              <strong>
                R$ 3,67
                 
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
